### ISSUE TITLE
Remove `repository` declarations from published pom

### ISSUE THREAD

**Issue by snicoll** [vaadin/platform#794]
Released artifacts have pom metadata with `repository` declaration. When that happens and such artifact is added as a dependency, Maven will use the declared repositories as a source to download all artifacts (not only the one related to the dependency). 

This is a major issue for users behind a corporate proxy as they don't want Maven to start hitting a service they don't have access to. I guess this would slow down things as I am expecting maven to reach out to get some metadata from each of them.

If I am adding the Spring Boot starter for the latest Vaadin 14 release, here is what I get by issuing `mvn dependency:list-repositories`:

```
INFO] --- maven-dependency-plugin:3.1.1:list-repositories (default-cli) @ demo ---
[INFO] Repositories used by this build:
[INFO]        id: sonatype-nexus-snapshots
      url: https://oss.sonatype.org/content/repositories/snapshots
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => false, update => daily]

[INFO]        id: vaadin-prerelease
      url: https://maven.vaadin.com/vaadin-prereleases
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => true, update => daily]

[INFO]        id: vaadin-addons
      url: https://maven.vaadin.com/vaadin-addons
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]

[INFO]        id: spring-milestones
      url: https://repo.spring.io/milestone
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]

[INFO]        id: vaadin-snapshots
      url: https://oss.sonatype.org/content/repositories/vaadin-snapshots/
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => false, update => daily]

[INFO]        id: apache.snapshots
      url: https://repository.apache.org/snapshots
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => false, update => daily]

[INFO]        id: central
      url: https://repo.maven.apache.org/maven2
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]

[INFO]        id: vaadin-prereleases
      url: https://maven.vaadin.com/vaadin-prereleases
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => true, update => daily]

```

You can reproduce by going on start.spring.io and add vaadin.

Here is the output if I remove the vaadin starter (replace that with `spring-boot-starter`).

```
[INFO] --- maven-dependency-plugin:3.1.1:list-repositories (default-cli) @ demo ---
[INFO] Repositories used by this build:
[INFO]        id: apache.snapshots
      url: https://repository.apache.org/snapshots
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => false, update => daily]

[INFO]        id: central
      url: https://repo.maven.apache.org/maven2
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]
```

We've identified the source for `apache.snapshots` and will follow-up with the concerned project.

A few observations:

* If you are using a GA release of Vaadin and snapshots for unrelated dependencies (your own, etc), Maven would hit https://oss.sonatype.org/content/repositories/snapshots and https://maven.vaadin.com/vaadin-prereleases and https://oss.sonatype.org/content/repositories/vaadin-snapshots/
* The https://maven.vaadin.com/vaadin-prereleases is present twice with (one with `s` at the end and one without)

I'd argue that adding the starter for a release shouldn't bring anything at all if the bits are in central which I believe is the case.

For the record, we use the `flatten-maven-plugin` to clean up poms before they are published. It allows us to remove build specific settings when publishing so that it doesn't pollute user's builds.

**Comment by juhopiirainen** [comment]
Hi and thanks for the suggestion!

We are looking into this and it seems promising. 
I'll leave this issues open and update with news.

**Comment by Legioth** [comment]
This is getting extra urgent at this moment since https://oss.sonatype.org has performance issues which makes that additional repository slow down things quite significantly.

**Comment by ZheSun88** [comment]
I have made a PR to the vaadin-parent project. 
https://github.com/vaadin/maven-integration/pull/10

added the mentioned `flatten-maven-plugin` to the project inside a separate profile. 
tested locally : 

with one of our released version of `vaadin-button-flow` dependency, we can see the following info
```
[INFO] --- maven-dependency-plugin:3.1.1:list-repositories (default-cli) @ demo                                                                                             ---
[INFO] Repositories used by this build:
[INFO]        id: vaadin-snapshots
      url: https://oss.sonatype.org/content/repositories/vaadin-snapshots/
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => true, update => daily]

[INFO]        id: central
      url: https://repo.maven.apache.org/maven2
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]

[INFO]        id: vaadin-prereleases
      url: https://maven.vaadin.com/vaadin-prereleases
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => true, update => daily]
```

and with flattened pom:
```
[INFO] --- maven-dependency-plugin:3.1.1:list-repositories (default-cli) @ demo ---
[INFO] Repositories used by this build:
[INFO]        id: central
      url: https://repo.maven.apache.org/maven2
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]
```

it looks okay, but we need to test with CI release, there might need some extra changes. 

@alvarezguille and @Legioth Mind to take a look?

**Comment by Legioth** [comment]
How do I test this with a platform release?

**Comment by ZheSun88** [comment]
An easy way to test this would be 

1. copy the plugin settings from https://github.com/vaadin/maven-integration/tree/add-flatten-maven-plugin and add it to a, let us say, component project (vaadin-button-flow) for example
2. build the component project with using the profile containing the plugin setting. 
3. replace the pom file with the flatten pom and build the project
4. take a skeleton project and remove all vaadin dependency, then add the new build component dependency
5 run the `mvn dependency:list-repositories`

for comparison, you can change the component version to a released one and run the same command.  
----------------------------------------------------------------------------------------------------------
the way that i have tested is a bit more complicated, as i edited the parent:

1. take https://github.com/vaadin/maven-integration/tree/add-flatten-maven-plugin and build `vaadin-parent` locally
2. take `flow-component-base` and update the parent project version to the installed one
3. run `mvn install -Pflatten-pom`
4. update to use the flattened pom (there should be a parameter to call directly, but i haven't found it yet.)
===============(you can test with `mvn dependency:list-repositories` already here, should be only central repo left. or you can do the following things to test a real project)
5. build the flow-component-base project, take `vaadin-button-flow` and update the parent project to the new build one
6. run `mvn install -Pflatten-pom`
7. manually update the pom to use the flatten one,  build the project. 
8. create an empty project without any dependencies on vaadin product.
5. add the `vaadin-button-flow` dependency and run `mvn dependency:list-repositories`

**Comment by Legioth** [comment]
The thing I'd like to verify is that the next Vaadin release won't contain the same `<repository>` definitions that can be seen in e.g. https://repo1.maven.org/maven2/com/vaadin/vaadin-platform-parent/14.0.5/vaadin-platform-parent-14.0.5.pom

**Comment by ZheSun88** [comment]
I would not do this test with an stable release, as it might contain some kind of expected result which will lead to a faulty release. Maybe

**Comment by Legioth** [comment]
So how can I know that the thing I test using your instructions will be the same thing that a regular user will get in an upcoming stable release?

The problem that I have observed is about contents of `vaadin-platform-parent`, whereas your instructions suggest doing special things when building e.g. `vaadin-button-flow`. 

I fail to understand how adding a Maven profile to one project will cause the contents of `pom.xml` in a different project to end up with different contents. Will the profile be activated for all builds of `vaadin-platform-parent` or only some of them?

**Comment by ZheSun88** [comment]
I think you have misunderstood the project name.. it is `vaadin-parent` , not `vaadin-platform-parent`

it is the parent project for all vaadin products. https://github.com/vaadin/platform/blob/master/pom.xml#L5...L10 

you can find the project here https://github.com/vaadin/maven-integration/tree/master/ .

**Comment by Legioth** [comment]
When you resolve dependencies for a typical Vaadin project, then it's the `<repository>` definitions in `vaadin-platform-parent` that first trigger Maven to start looking up dependencies from oss.sonatype.org. That's why I'm concerned about that artefact. It might not be the only source of trouble, but it is the first one that is encountered.

**Comment by ZheSun88** [comment]
yeah, i get your concern. Made a PR to change/correct the order of repositories which are defined in the POM. https://github.com/vaadin/platform/pull/891

maven central > prerelease >snapshot.

but this is a bit different with this original issue in this ticket, as we still bring in prerelease and some other unnecessary repository dependencies in for the user.  But let us see how it go with the corrected repository order.

**Comment by snicoll** [comment]
Do I understand you do not intend to remove the `<repositories>` and `<pluginRepositories>` from the platform pom?

**Comment by ZheSun88** [comment]
@snicoll no no.. we will.. 

@Legioth is just pointing to another issue with us.

**Comment by ZheSun88** [comment]
PR for adding `flatten-maven-plugin` has been merged and released with `vaadin-parent  2.0.2`

we have tried with **platform 14.2.0.alpha8** release, but as repository declarations are still defined in platform's components, it cannot solve all the issue mentioned here at once. 

next step, we need to update all the components (flow/flow-components, addon projects) to use the new `vaadin-parent`

**Comment by ZheSun88** [comment]
tested with a starter project with `16.0.0.rc1` 
project defines `central`, `vaadin-prereleaes` and `vaadin-directory`. 

```
[INFO] Repositories Used by this build:
[INFO]       id: central
      url: https://repo.maven.apache.org/maven2
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]
[INFO]       id: Vaadin Directory
      url: https://maven.vaadin.com/vaadin-addons
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]
[INFO]       id: sonatype-snapshots
      url: https://oss.sonatype.org/content/repositories/snapshots
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => true, update => daily]
[INFO]       id: sonatype-releases
      url: https://oss.sonatype.org/content/repositories/releases
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]
[INFO]       id: sonatype-nexus-snapshots
      url: https://oss.sonatype.org/content/repositories/snapshots
   layout: default
snapshots: [enabled => true, update => daily]
 releases: [enabled => false, update => daily]
[INFO]       id: vaadin-prereleases
      url: https://maven.vaadin.com/vaadin-prereleases/
   layout: default
snapshots: [enabled => false, update => daily]
 releases: [enabled => true, update => daily]
```

the `sonatype-x` repos are coming from dependencies used in our product for example our own forks from `gentyref` ([HERE](https://github.com/vaadin/gentyref/blob/master/pom.xml#L24...L35 )) and probably other 3rd party dependencies too. 

we have fixed the main issue with 
- same repo with different repo IDs
- totally removed the `vaadin-snapshots` usage.
